From 16b1f282270441b956c0fc5279d1036e5822b28a Mon Sep 17 00:00:00 2001
From: Horia Geanta <horia.geanta@freescale.com>
Date: Fri, 18 Apr 2014 13:01:42 +0300
Subject: [PATCH 0875/5242] crypto: caam - add allocation failure handling in
 SPRINTFCAT macro

commit  640e67fd821edb56ae0f1b05f4881dbd51125d90 from
https://source.codeaurora.org/external/imx/linux-imx.git

commit 27c5fb7a84242b66bf1e0b2fe6bf40d19bcc5c04 upstream.

GFP_ATOMIC memory allocation could fail.
In this case, avoid NULL pointer dereference and notify user.

Cc: Kim Phillips <kim.phillips@freescale.com>
Signed-off-by: Horia Geanta <horia.geanta@freescale.com>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Conflicts:
	drivers/crypto/caam/error.c
Signed-off-by: Dan Douglass <dan.douglass@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/caam/error.c |   14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/crypto/caam/error.c b/drivers/crypto/caam/error.c
index 8da88be..8c37a3c 100644
--- a/drivers/crypto/caam/error.c
+++ b/drivers/crypto/caam/error.c
@@ -156,6 +156,20 @@ void caam_dump_sg(const char *level, const char *prefix_str, int prefix_type,
 	"Secure key generation",
 };
 
+#define SPRINTFCAT(str, format, param, max_alloc)		\
+{								\
+	char *tmp;						\
+								\
+	tmp = kmalloc(sizeof(format) + max_alloc, GFP_ATOMIC);	\
+	if (likely(tmp)) {					\
+		sprintf(tmp, format, param);			\
+		strcat(str, tmp);				\
+		kfree(tmp);					\
+	} else {						\
+		strcat(str, "kmalloc failure in SPRINTFCAT");	\
+	}							\
+}
+
 static void report_ccb_status(struct device *jrdev, const u32 status,
 			      const char *error)
 {
-- 
1.7.9.5

